### PR TITLE
log_event_decoder: time: plug 2038 problem on timestamp

### LINKED ISSUE #11532: Fluentbit issue when using lua script after 2038
## Bug Report

**Describe the bug**
We use fluentbit and lua script without any issue. But when testing it after 2038 it does not work anymore with an error linked to a 32 bits variable

**To Reproduce**
- Example log message if applicable:
```
[2040/01/03 03:15:10.808241446] [debug] [decoder] Invalid group marker timestamp (-2085883291), skipping record. Group state preserved.

```
- Steps to reproduce the problem:
Configure FLuentbit using a lua script. Change the date to 2040

**Expected behavior**
Fluentbit should handle timestamp in 64 bits to work after 2038

**Screenshots**

**Your Environment**

* Version used: Fluentbit vers 4.2.3
* Configuration:  Use of LUA script
* Environment name and version (e.g. Kubernetes? What version?): Hosted in k3S
* Server type and version: 
* Operating System and version: Host OS is SLES 16.0 in 64 bits
* Filters and plugins:

**Additional context**
We want to have a solution working after the bug 2038

### PR DESCRIPTION
**Pull request by cosmo0920**
Fixes https://github.com/fluent/fluent-bit/issues/11532.

On Fluent Bit core, we still have 2038 issues.
second part of flb_tm struct was processed as int32_t type and out_stdout has int32_t assumed part of printing timestamps.
So, those of operations can be truncate after 2038 cut-off of timestamps issue.
We need to fix and process after surpassed of 2038 timestamps.

After patching this, we can process until 2106 A.D logs.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:

- [x] Example configuration file for the change
```conf
[SERVICE]
    Flush        1
    Daemon       Off
    Log_Level    debug

[INPUT]
    Name         dummy
    Tag          test.lua
    Dummy        {"KEY_OF_TIMESTAMP":2209161600.123456,"msg":"hello-from-2040"}
    Samples      1

[FILTER]
    Name         lua
    Match        test.lua
    Script       override_time_2040.lua
    Call         override_time_2040

[OUTPUT]
    Name         stdout
    Match        test.lua
```

* override_time_2040.lua
```lua
function override_time_2040(tag, timestamp, record)
    return 1, record["KEY_OF_TIMESTAMP"], record
end
```
- [x] Debug log output from testing the change
```
% bin/fluent-bit -c future-of-lua-script.conf -v
Fluent Bit v5.0.2
* Copyright (C) 2015-2026 The Fluent Bit Authors
* Fluent Bit is a CNCF graduated project under the Fluent organization
* https://fluentbit.io

______ _                  _    ______ _ _           _____  _____ 
|  ___| |                | |   | ___ (_) |         |  ___||  _  |
| |_  | |_   _  ___ _ __ | |_  | |_/ /_| |_  __   _|___ \ | |/' |
|  _| | | | | |/ _ \ '_ \| __| | ___ \ | __| \ \ / /   \ \|  /| |
| |   | | |_| |  __/ | | | |_  | |_/ / | |_   \ V //\__/ /\ |_/ /
\_|   |_|\__,_|\___|_| |_|\__| \____/|_|\__|   \_/ \____(_)\___/


[2026/04/03 19:45:16.122] [ info] Configuration:
[2026/04/03 19:45:16.124] [ info]  flush time     | 1.000000 seconds
[2026/04/03 19:45:16.124] [ info]  grace          | 5 seconds
[2026/04/03 19:45:16.124] [ info]  daemon         | 0
[2026/04/03 19:45:16.124] [ info] ___________
[2026/04/03 19:45:16.124] [ info]  inputs:
[2026/04/03 19:45:16.124] [ info]      dummy
[2026/04/03 19:45:16.124] [ info] ___________
[2026/04/03 19:45:16.124] [ info]  filters:
[2026/04/03 19:45:16.124] [ info]      lua.0
[2026/04/03 19:45:16.124] [ info] ___________
[2026/04/03 19:45:16.124] [ info]  outputs:
[2026/04/03 19:45:16.124] [ info]      stdout.0
[2026/04/03 19:45:16.124] [ info] ___________
[2026/04/03 19:45:16.124] [ info]  collectors:
[2026/04/03 19:45:16.124] [ info] [fluent bit] version=5.0.2, commit=d835ea9ee9, pid=77022
[2026/04/03 19:45:16.124] [debug] [engine] coroutine stack size: 36864 bytes (36.0K)
[2026/04/03 19:45:16.124] [ info] [storage] ver=1.5.4, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2026/04/03 19:45:16.124] [ info] [simd    ] disabled
[2026/04/03 19:45:16.124] [ info] [cmetrics] version=2.1.1
[2026/04/03 19:45:16.125] [ info] [ctraces ] version=0.7.1
[2026/04/03 19:45:16.125] [ info] [input:dummy:dummy.0] initializing
[2026/04/03 19:45:16.125] [ info] [input:dummy:dummy.0] storage_strategy='memory' (memory only)
[2026/04/03 19:45:16.125] [debug] [dummy:dummy.0] created event channels: read=21 write=22
[2026/04/03 19:45:16.125] [debug] [stdout:stdout.0] created event channels: read=23 write=24
[2026/04/03 19:45:16.125] [ info] [sp] stream processor started
[2026/04/03 19:45:16.125] [ info] [output:stdout:stdout.0] worker #0 started
[2026/04/03 19:45:16.126] [ info] [engine] Shutdown Grace Period=5, Shutdown Input Grace Period=2
[2026/04/03 19:45:18.129] [debug] [task] created task=0x6000012b4000 id=0 OK
[2026/04/03 19:45:18.129] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] test.lua: [[2209161600.123456001, {}], {"KEY_OF_TIMESTAMP"=>2209161600.123456, "msg"=>"hello-from-2040"}]
[2026/04/03 19:45:18.130] [debug] [out flush] cb_destroy coro_id=0
[2026/04/03 19:45:18.130] [debug] [task] destroy task=0x6000012b4000 (task_id=0)
^C[2026/04/03 19:45:32] [engine] caught signal (SIGINT)
[2026/04/03 19:45:32.865] [ info] [input] pausing dummy.0
[2026/04/03 19:45:32.865] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2026/04/03 19:45:32.865] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

It's not valgrind but it ran under leaks command on macOS:

```
Process 77174 is not debuggable. Due to security restrictions, leaks can only show or save contents of readonly memory of restricted processes.

Process:         fluent-bit [77174]
Path:            /Users/USER/*/fluent-bit
Load Address:    0x10205c000
Identifier:      fluent-bit
Version:         0
Code Type:       ARM64
Platform:        macOS
Parent Process:  leaks [77173]
Target Type:     live task

Date/Time:       2026-04-03 19:46:03.932 +0900
Launch Time:     2026-04-03 19:45:51.097 +0900
OS Version:      macOS 15.7.4 (24G517)
Report Version:  7
Analysis Tool:   /usr/bin/leaks

Physical footprint:         4737K
Physical footprint (peak):  4881K
Idle exit:                  untracked
----

leaks Report Version: 4.0, multi-line stacks
Process 77174: 750 nodes malloced for 81 KB
Process 77174: 0 leaks for 0 total leaked bytes.

[2026/04/03 19:46:04] [engine] caught signal (SIGCONT)
----

leaks Report Version: 4.0, multi-line stacks
Process 77148: 750 nodes malloced for 81 KB
Process 77148: 0 leaks for 0 total leaked bytes.
```

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.

- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [ ] Documentation required for this feature

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed timestamp handling to correctly support dates beyond 2038.
  * Improved decoding of extended timestamp formats to preserve group markers and edge cases.
  * Updated timestamp output formatting for accurate 64-bit second values across components.
  * Strengthened validation and logging for malformed or special-case timestamps.
  * Added/updated tests to verify post-2038 timestamp behavior.

### CODE REVIEW

File: src/flb_log_event_decoder.c [hunk 1]
```diff
@@ -190,19 +192,42 @@ int flb_log_event_decoder_decode_timestamp(msgpack_object *input,
...
+         * -1000), which are encoded near UINT32_MAX.
+         */
+        if (sec == (uint32_t) -1) {
+            output->tm.tv_sec = FLB_LOG_EVENT_GROUP_START;
+        }
+        else if (sec == (uint32_t) -2) {
+            output->tm.tv_sec = FLB_LOG_EVENT_GROUP_END;
+        }
+        else if (sec >= 0xFFFFFC00U) {
+            output->tm.tv_sec = (time_t) ((int32_t) sec);
```
[the remark this answers is not in the document]
Comment (cosmo0920) [reply src/flb_log_event_decoder.c]: Yeah, but this limitation is needed to represent invalid timestamps from -1 to -1024.
This behavior is intentional and documented here:

/*
 * EventTime stores seconds as unsigned 32-bit.
 * Keep -1/-2 semantics for group markers, allow normal post-2038
 * timestamps, and preserve handling for intentionally crafted negative
 * marker-like values used by tests/corruption checks (e.g. -3, -10,
 * -1000), which are encoded near UINT32_MAX.
 */

So we consider this restriction valid in our implementation.

**Comment by patrick-stephens** [comment]
@cosmo0920 I'd probably say let's do a build of all targets just to confirm no problems anywhere but I think it's fine.

**Comment by cosmo0920** [comment]
This branch is still underdevelopment so I removed ok-package-test for now.

**Comment by edsiper** [comment]
Moving it to next milestone (pls confirm once is ready)

**Comment by dirkmueller** [comment]
@cosmo0920 anything I can help with?

**Review by dirkmueller — APPROVED** [review]
LGTM. great work.

**Comment by mlaguilharre** [comment]
@cosmo0920 kindly inform that though 2038 seems far, we could face legal action if we provide today code that can't cope with 2038 date. Could you help us to move forward or let us know if we can help.
